lib/dogma/rule/taken_name.ex

```diff
@@ -21,59 +21,102 @@ defrule Dogma.Rule.TakenName do
     unquote_splicing update_in use
   )a
 
-  @reserved_words Enum.into(reserved_words, MapSet.new)
+  @other_reserved_words Enum.into(other_reserved_words, MapSet.new)
+
+  @protocol_with_reserved_words %{
+    [:Inspect] => :inspect,
+    [:List, :Chars] => :to_charlist,
+    [:String, :Chars] => :to_string
+  }
 
   @spec reserved_words :: MapSet.t
   def reserved_words do
-    @reserved_words
+    @protocol_with_reserved_words
+    |> Map.values()
+    |> Enum.into(@other_reserved_words)
+  end
+
+  @spec protocol_with_reserved_words :: %{required([atom]) => atom}
+  def protocol_with_reserved_words do
+    @protocol_with_reserved_words
   end
 
   def test(_rule, script) do
-    script |> Script.walk( &check_ast(&1, &2) )
+
+    {_, {_, errors}} =
+      Macro.traverse( script.ast, {reserved_words(), []},
+        &check_ast(&1, &2),
+        &end_of_impl(&1, &2))
+    errors
   end
 
-  defp check_ast({:def, _, [{name, meta, _} | _]} = ast, errors) do
-    check_name(name, meta, ast, errors)
+  defp check_ast({:def, _, [{name, meta, _} | _]} = ast, acc) do
+    check_name(name, meta, ast, acc)
+  end
+  defp check_ast({:defmacro, _, [{name, meta, _} | _]} = ast, acc) do
+    check_name(name, meta, ast, acc)
+  end
+  defp check_ast({:defp, _, [{name, meta, _} | _]} = ast, acc) do
+    check_name(name, meta, ast, acc)
   end
-  defp check_ast({:defmacro, _, [{name, meta, _} | _]} = ast, errors) do
-    check_name(name, meta, ast, errors)
+  defp check_ast(
+    {:defimpl, _, [{_, _, impl_name} | _]} = ast,
+    {reserved, errors}) do
+    new_reserved = maybe_alter_words(reserved, impl_name, &MapSet.delete/2)
+    {ast, {new_reserved, errors}}
   end
-  defp check_ast({:defp, _, [{name, meta, _} | _]} = ast, errors) do
-    check_name(name, meta, ast, errors)
+  defp check_ast(ast, acc) do
+    {ast, acc}
   end
-  defp check_ast(ast, errors) do
-    {ast, errors}
+
+  defp end_of_impl(
+    {:defimpl, _, [{_, _, impl_name} | _]} = ast,
+    {reserved, errors}) do
+    new_reserved = maybe_alter_words(reserved, impl_name, &MapSet.delete/2)
+    {ast, {new_reserved, errors}}
+  end
+  defp end_of_impl(ast, acc) do
+    {ast, acc}
   end
 
-  defp valid_name?(name) do
-    ! MapSet.member?(@reserved_words, name)
+  defp valid_name?(name, reserved) do
+    ! MapSet.member?(reserved, name)
   end
 
-  defp check_name(name, meta, ast, errors) when is_atom(name) do
-    if valid_name?(name) do
-      {ast, errors}
+  defp check_name(name, meta, ast, {reserved, errors}) when is_atom(name) do
+    if valid_name?(name, reserved) do
+      {ast, {reserved, errors}}
     else
-      {ast, [error( meta[:line], name ) | errors]}
+      {ast, {reserved, [error( meta[:line], name ) | errors]}}
     end
   end
 
-  defp check_name(_, _, ast, errors) do
-    {ast, errors}
+  defp check_name(_, _, ast, acc) do
+    {ast, acc}
+  end
+
+  defp maybe_alter_words(reserved, impl_name, fun) do
+    if Map.has_key?(protocol_with_reserved_words(), impl_name) do
```

Comment:
Rather than looking in the map twice why not use `Map.fetch` and pattern matching on the results with `fetch`? :)